libusb/os/windows_winusb.c

```diff
@@ -2293,6 +2293,7 @@ const struct windows_backend winusb_backend = {
 	winusb_cancel_transfer,
 	winusb_clear_transfer_priv,
 	winusb_copy_transfer_data,
+	"WinUSB",
```

Comment:
Thank you for the PR!
I think here I am not agreeing with you that this is the driver name. As far as I understand, UsbDK and WinUSB are 2 different Windows backenda. The "driver" as per Windows meaning is something else.
For WinUSB backed it can be WinUSB, but also libusbK or libusb0 (or something new in the future).

When the feature was presented in issue #1225 it clearly stated that for Windows the new API shall return one of those 3 for now, see: https://github.com/libusb/libusb/issues/1225#issue-1514927539.

Therefore I think that here we should have a function instead of a fixed string, in order to return actually loaded driver. Otherwise the feature makes less sense to me.

Thanks for considering this.

Reply:
> As far as I understand, UsbDK and WinUSB are 2 different Windows backend

Agreed.

> For WinUSB backed it can be WinUSB, but also libusbK or libusb0 (or something new in the future).

Agreed but there are actually three types, resulting in eights APIs which can be tapped into:

- Composite devices: dg_ssudbus USBCCGP,
- WinUsbX: libusbK, libusb0, WinUSB
- HID: HIDUSB, MOUHID, KBDHID

However, these are *per device* APIs. So we can't return either of these names when driver name is requested for a backend because it could be any of them.

> for Windows the new API shall return one of those 3 for now, see: https://github.com/libusb/libusb/issues/1225#issue-1514927539.

I don't think this list is accurate. 

- I think `libusbK` is a typo. The author probably meant `libusbdK`. 
- There is no `libusb-win32` support. We have pieces of code from `libusb-win32` but that is neither a backend nor a supported API. 
- As for `WinUSB`, as you pointed, out it is a bad name. I took it from the translation unit but I shouldn't have.

My proposal for Windows is to keep `libusbdK` for the `windows_backend` defined in `windows_usbdk.c`. And use `Per-Device-API` (or `Composite-WinUsbX-HID`)  for the `windows_backend` defined in `windows_winusb.c`

Reply:
Hello @fabiensanglard , thank you for your prompt reply. Yes, you're right, I wrote my comment too quickly: given the function prototype you chose (and correctly advertised in #1225 before doing your development, you cannot really return a driver name, since it only make sens per device.
However, given this "detail" that I didn't realize earlier, I am now in doubt about what additional information this work brings. It is more like getting the backend name (which is useful per se) and / or some hardcoded static driver name that is 1:1 with backend name. 

So, I would suggest that either rename the function to get_backend_name or add a "libusb_device_handle *" parameter to the current get_driver_name function and implement it for returning the drriver name under Windows.
But we need @mcuee opinion here as he's the one who first requested the feature.

Regarding libusbK: it really is libusK (not libusbdK): https://github.com/mcuee/libusbk
Regarding libusb-win32 aka libusb0.sys: it is supported via libusk.dll (not a driver but an user space lib) that interfaces in between libusb-1.0 and libusb0.sys (it provide a "WinUBS-like" API over libusb0.sys .

So well, @mcuee : do we want backend name or driver name (or both) ?

Reply:
My original intention: just the driver name and not the backend, as there is no extra info given by the backend.

Windows WinUSBx backend -- return `WinUSB` (WinUSB.dll --> WinUSB.sys), `libusbK` (libusbK.dll --> libusbK.sys) or `libusb-win32` (libusbK.dll --. libusb0.sys) accordingly.

Windows USB HID backend -- return `hidusb`. As for USB HID Mouse and Keyboard, I am not so sure.  libusb Windows can not do much with USB HID Mouse/Keyboard.

Windows USBDK backend -- return `usbdk`

Linux -- return `usbfs`
macOS -- return `IOKit`
BSDs -- return `ugen`

USB Composite Device is a tricky part. It was not really my intention to return `USBCCGP` (or `dg_ssudbus` as a special case for Samsung devices).

Reply:
It was more meaningful to get the backend name last time, at least for Windows.

Last time, it was possible to access devices using WinUSB driver through two ways.
1) No libusbK.dll present
`libusb-1.0.dll --> WinUSB.dll --> WinUSB.sys`
2) With libusbK.dll present
`libusb-1.0.dll --> libusbK.dll --> WinUSB.dll --> WinUSB.sys`

Now there is only one way, no matter the presence of libusbK.dll.
`libusb-1.0.dll --> WinUSB.dll --> WinUSB.sys`

Reply:
But after reading the info, maybe it is worth getting the backend name as well in the end to have more flexibilities. In the future, maybe there will be cases that there will be two ways to access a USB device (same device driver) using different backends.

Reply:
Let's hear from @hjelmn and @tormodvolden as well for their opinions.

Reply:
Based on both your feedback (@mcuee and @somatique) it seems what is needed is the driver name **per device**. For that I need to change the proposed API to:

```
const char* libusb_get_driver_name(struct libusb_device *device);
```

If you are both ok, I can update the CL.

Reply:
> 

> const char* libusb_get_driver_name(struct libusb_device *device);
In my opinion this is the best way to go.

> It would make sense to leverage the names in `composite_driver_names`, `winusbx_driver_names`, and `hid_driver_names` since they are already in the `windows_usb_api_backend`. If you don't like the values, maybe we can change them?

I think current values are fine.
Thanks!

Reply:
> If you are both ok, I can update the CL.

Okay, please proceed. Thanks.

Hopefully @tormodvolden or @hjelmn would have some time later to review the proposal as well.